AX: [AX Thread Text APIs] Center-aligned text has incorrect frames

hoffmanjoshua opened the pull request:
#### ef09367a65fab81d60ee70f2e5c98c46077fa145
<pre>
AX: [AX Thread Text APIs] Center-aligned text has incorrect frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=292295">https://bugs.webkit.org/show_bug.cgi?id=292295</a>
<a href="https://rdar.apple.com/149509105">rdar://149509105</a>

Reviewed by Tyler Wilcock.

When a run in text that is center-aligned is smaller than its containing RenderText,
we were not positioning its bounding box correctly. Our algorithm failed to consider
the inset from the render text&apos;s frame to the run itself, leading to left-aligned
frames for centered text.

* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::AXTextRun):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):

Computes and caches a new property, distanceFromBoundsInDirection, on each text run.
This property represents the distance between the RenderText&apos;s frame and the start of
the run (in either the horizontal or vertical direction). To compute this, we cache
the offset properties provided by the text run (see comments in code) to compute the
distance from the containing block, the render texts element rect, and the rect of the
containing block.

* LayoutTests/accessibility/mac/bounds-for-range-centered-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-centered.html: Added.
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(parameterizedAttributeValueForTesting):
(computeTextBoundsForRange):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
(WTR::AccessibilityUIElement::boundsForRangeWithPagePosition):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::makeBoundsDescription):
(WTR::AccessibilityUIElement::boundsForRange):
(WTR::AccessibilityUIElement::boundsForRangeWithPagePosition):

New testing infrastructure to allow text bounds to be returned which are relative to
the page, not the screen. This makes for more consistent and reasonable numbers during
layout tests.

Canonical link: <a href="https://commits.webkit.org/294340@main">https://commits.webkit.org/294340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e4bf365cbfbe62ee340c58332e3a3ab46dad109

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21110 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106597 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52073 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103484 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29602 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77259 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34287 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104451 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91618 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57601 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16346 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9635 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51421 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86221 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9708 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108949 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21019 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86231 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28935 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87824 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85793 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21840 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30526 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8239 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22690 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28504 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33784 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28315 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31635 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29874 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextRun.h:
```diff
@@ -83,12 +83,16 @@ struct AXTextRun {
     float lineHeight;
 
-    AXTextRun(size_t lineIndex, String&& text, Vector<std::array<uint16_t, 2>>&& domOffsets, Vector<uint16_t> characterAdvances, float lineHeight)
+    // The distance between the RenderText's position and the start of the text run (useful for things that are not left-aligned).
```
I feel like in a year from now we'll read "useful for things that are not left-aligned" and think "like what" 🙂 With this problem fresh in our mind, it's obvious that this refers to `text-align: center`, but IMO this reads just a bit vague enough that we won't remember in the future. 

Would it make sense to specifically call out `text-align: center` as one example of where this is relevant?

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1456,6 +1457,20 @@ AXTextRuns AccessibilityRenderObject::textRuns()
+        if (CheckedPtr containingBlock = renderText->containingBlock())
+            containingBlockOffset = fontOrientation() == FontOrientation::Horizontal ? containingBlock->absoluteBoundingBoxRect().x() : containingBlock->absoluteBoundingBoxRect().y();
+
+        distanceFromBoundsInDirection = fontOrientation() == FontOrientation::Horizontal ? textBox->lineBox()->contentLogicalLeft() + containingBlockOffset - elementRect().x() : -textRun.xPos() + containingBlockOffset - elementRect().y();
```
We're adding new calls to `absoluteBoundingBoxRect` and `elementRect`, both of which I believe are extremely expensive. Do you think you could measure the impact of this change on a page with a lot of text, like https://html.spec.whatwg.org, so we know what effect this has on performance?

Also, `fontOrientation()` requires some virtual function calls, so rather than re-computing it multiple times, could we compute it once and store it in a local `isHorizontal` variable or something like that?

hoffmanjoshua replied:
Gave the HTML spec page a whirl and determined that this still doesn't add any meaningful complexity to this page based on my sample. The page still takes a while to build all text runs, but for unrelated reasons. So shouldn't be a concern.

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3222,7 +3222,7 @@ - (id)accessibilityAttributeValue:(NSString*)attribute forParameter:(id)paramete
     }
 
-    if ([attribute isEqualToString:NSAccessibilityBoundsForRangeParameterizedAttribute]) {
+    if ([attribute isEqualToString:NSAccessibilityBoundsForRangeParameterizedAttribute] || [attribute isEqualToString:_AXPageBoundsForTextMarkerRangeAttribute]) {
```
I don't think we should handle this private, testing-only attribute in this function. Instead, we should probably handle it in `parameterizedAttributeValueForTesting`

hoffmanjoshua replied:
Moved this to the testing method, and abstracted out the logic to `computeTextBoundsForRange` so it can be shared across the two places we need it.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1456,6 +1457,20 @@ AXTextRuns AccessibilityRenderObject::textRuns()
+        if (CheckedPtr containingBlock = renderText->containingBlock())
+            containingBlockOffset = fontOrientation() == FontOrientation::Horizontal ? containingBlock->absoluteBoundingBoxRect().x() : containingBlock->absoluteBoundingBoxRect().y();
+
+        distanceFromBoundsInDirection = fontOrientation() == FontOrientation::Horizontal ? textBox->lineBox()->contentLogicalLeft() + containingBlockOffset - elementRect().x() : -textRun.xPos() + containingBlockOffset - elementRect().y();
```
Is it safe to unconditionally dereference `textBox->lineBox()`?

hoffmanjoshua replied:
It should be (in many places where `textBox->lineBox()` is used, it is dereferenced immediately. But, there's no harm in being extra safe here so I added a null check, and moved a preexisting line that did this behind that condition.

hoffmanjoshua committed:
AX: [AX Thread Text APIs] Center-aligned text has incorrect frames
https://bugs.webkit.org/show_bug.cgi?id=292295
rdar://149509105

Reviewed by Tyler Wilcock.

When a run in text that is center-aligned is smaller than its containing RenderText,
we were not positioning its bounding box correctly. Our algorithm failed to consider
the inset from the render text's frame to the run itself, leading to left-aligned
frames for centered text.

* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::AXTextRun):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):

Computes and caches a new property, distanceFromBoundsInDirection, on each text run.
This property represents the distance between the RenderText's frame and the start of
the run (in either the horizontal or vertical direction). To compute this, we cache
the offset properties provided by the text run (see comments in code) to compute the
distance from the containing block, the render texts element rect, and the rect of the
containing block.

* LayoutTests/accessibility/mac/bounds-for-range-centered-expected.txt: Added.
* LayoutTests/accessibility/mac/bounds-for-range-centered.html: Added.
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(parameterizedAttributeValueForTesting):
(computeTextBoundsForRange):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
(WTR::AccessibilityUIElement::boundsForRangeWithPagePosition):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::makeBoundsDescription):
(WTR::AccessibilityUIElement::boundsForRange):
(WTR::AccessibilityUIElement::boundsForRangeWithPagePosition):

New testing infrastructure to allow text bounds to be returned which are relative to
the page, not the screen. This makes for more consistent and reasonable numbers during
layout tests.

Canonical link: https://commits.webkit.org/294340@main
